samples/cpp/tutorial_code/gapi/security_barrier_camera/security_barrier_camera.cpp

```diff
@@ -0,0 +1,351 @@
+    }.cfgOutputLayers({ "color", "type" });
+
+    // Fill a special LPR input (seq_ind) with a predefined value
+    // First element is 0.f, the rest 87 are 1.f
+    const std::vector<int> lpr_seq_dims = {88,1};
+    cv::Mat lpr_seq(lpr_seq_dims, CV_32F, cv::Scalar(1.f));
+    lpr_seq.ptr<float>()[0] = 0.f;
+    auto lpr_net = cv::gapi::ie::Params<custom::LPR> {
+        cmd.get<std::string>("lprm"),   // path to topology IR
+        cmd.get<std::string>("lprw"),   // path to weights
+        cmd.get<std::string>("lprd"),   // device specifier
+    }.constInput("seq_ind", lpr_seq);
+
+    auto kernels = cv::gapi::kernels<custom::OCVProcessDetections>();
+    auto networks = cv::gapi::networks(det_net, attr_net, lpr_net);
+
+    Avg avg;
+    cv::Mat frame;
+    std::vector<cv::Rect> vehicles, plates;
+    std::vector<cv::Mat> out_colors;
+    std::vector<cv::Mat> out_types;
+    std::vector<cv::Mat> out_numbers;
+    std::size_t frames = 0u;
+
+    std::cout << "Reading " << input << std::endl;
+
+    if (cmd.get<bool>("ser")) {
+        std::cout << "Going serial..." << std::endl;
+        cv::VideoCapture cap(input);
+
+        auto cc = pp.compile(cv::GMatDesc{CV_8U,3,cv::Size(1920,1080)},
+                             cv::compile_args(kernels, networks));
+
+        avg.start();
+        while (cv::waitKey(1) < 0) {
+            cap >> frame;
+            if (frame.empty()) break;
+
+            cc(cv::gin(frame),
+               cv::gout(frame, vehicles, out_colors, out_types, plates, out_numbers));
+            frames++;
+            labels::DrawResults(frame, vehicles, out_colors, out_types, plates, out_numbers);
+            labels::DrawFPS(frame, frames, avg.fps(frames));
+            if (!no_show) cv::imshow("Out", frame);
+        }
+    } else {
+        std::cout << "Going pipelined..." << std::endl;
+
+        auto cc = pp.compileStreaming(cv::GMatDesc{CV_8U,3,cv::Size(1920,1080)},
+                                      cv::compile_args(kernels, networks));
+
+        cc.setSource(cv::gapi::wip::make_src<cv::gapi::wip::GCaptureSource>(input));
+
+        avg.start();
+        cc.start();
+
+        // Implement different execution policies depending on the display option
+        // for the best performance.
+        while (cc.running()) {
+            auto out_vector = cv::gout(frame, vehicles, out_colors, out_types, plates, out_numbers);
+            if (no_show) {
+                // This is purely a video processing. No need to balance with UI rendering.
+                // Use a blocking pull() to obtain data. Break the loop if the stream is over.
+                if (!cc.pull(std::move(out_vector)))
+                    break;
+            } else if (!cc.try_pull(std::move(out_vector))) {
+                // Use a non-blocking try_pull() to obtain data.
+                // If there's no data, let UI refresh (and handle keypress)
+                if (cv::waitKey(1) >= 0) break;
```

Comment:
It is tricky to call `waitKey()` before `imshow()`, especially under condition.

If `cc.try_pull(std::move(out_vector))` is true, then UI is not updated.

```
else if (needUpdateUI)
{
    needUpdateUI = false;
    waitKey... with break only
}
else if (!cc.try_pull(std::move(out_vector))) // or blocking .pull(), but with disabled keys handling
{
    needUpdateUI = false;
    waitKey... with break/continue
}

...

imshow(...);
needUpdateUI = true;
```

Reply:
I faced several inefficiencies when displaying every frame I get - compared to GStreamer's displaysink. The trick here is to pull the pipeline for data in an endless loop until the available data ends, and only then to display the latest acquired frame and spend some cycles in UI refresh. It is purely a trick, but helped to improve the overall FPS somehow